code/game/objects/machinery/field_charger.dm

```diff
@@ -0,0 +1,292 @@
+		return TRUE
+
+///Use whenever adding a cell to the charger to ensure things are properly updated
+/obj/machinery/power/field_charger/proc/add_cell_to_storage(cell)
+	LAZYADD(stored_cells, cell)
+	build_ui_list()
+	start_processing()
+
+/obj/machinery/power/field_charger/wrench_act(mob/living/user, obj/item/I)
+	setAnchored(!anchored)
+	balloon_alert(user, anchored ? "anchored" : "unanchored")
+	playsound(get_turf(src), 'sound/items/ratchet.ogg', 25, 1)
+
+/obj/machinery/power/field_charger/setAnchored(anchorvalue)
+	. = ..()
+	use_power = anchorvalue	//auto_use_power() doesn't need to be called if the machine isn't anchored
+	if(anchorvalue)
+		connect_to_network()
+	else
+		disconnect_from_network()
+
+/obj/machinery/power/field_charger/welder_act(mob/living/user, obj/item/I)
+	welder_repair_act(user, I, max_integrity/10)
+
+/obj/machinery/power/field_charger/ui_interact(mob/user, datum/tgui/ui)
+	if(!is_operational())
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
+
+	if(!ui)
+		ui = new(user, src, "FieldCharger", name)
+		ui.open()
+
+/obj/machinery/power/field_charger/ui_data(mob/user)
+	. = list()
+	//Need to copy the list because it might get updated while iterating
+	//Must loop through stored_item_references to copy each associative list
+	//Otherwise BYOND will just create references to the sublists instead of making new ones
+	var/list/formatted_list = list()
+	LAZYINITLIST(stored_item_references)
+	for(var/list/sublist in stored_item_references)
+		//Transform the "charge" entries from cell objects to their battery levels
+		var/obj/item/cell/cell = sublist["charge"]
+		if(!cell)	//If a cell doesn't exist, kick this item out
+			eject_item(sublist["ref"])
+			continue
+
+		var/list/copied_sublist = sublist.Copy()
+		copied_sublist["charge"] = cell.percent()
+		formatted_list += list(copied_sublist)
+
+	//Return the list of items
+	.["items"] = formatted_list
+	//Send the UI the charger's battery level
+	.["charge"] = PERCENT(machine_current_charge / machine_max_charge)
+	.["connected"] = powernet ? TRUE : FALSE
+
+/obj/machinery/power/field_charger/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+
+	if(!is_operational())
+		return
+
+	//Best to use switch(action) but we're only looking for one type of action: the eject button
+	if(action == "eject")
+		eject_item(params["eject"], usr)
+		. = TRUE
+
+///Handles dispensing items from the charger; reference_to_item MUST be a reference, user is optional
+/obj/machinery/power/field_charger/proc/eject_item(reference_to_item, mob/user)
+	//Look for the item in the stored references list; only way to convert from a reference back to an item
+	var/atom/movable/item
+	for(var/list/list AS in stored_item_references)
+		var/atom/item_entry = list["ref"]
+		if(!item_entry)	//Maybe the item got deleted? If so, dispose of that entry
+			LAZYREMOVE(stored_item_references, list)
+			continue
+
+		if(item_entry == reference_to_item)
+			item = locate(item_entry) in contents
+
```

Comment:
Idk what all this is but why not just use locate() on the string ref and check it's in the list instead of all this string ref key handling?

Reply:
I didn't know I could do that. Neat.